### PR TITLE
PF1664: reference cloning instructions

### PR DESCRIPTION
**Pull request by ginaygoogle**

### CODE REVIEW

File: src/main/java/bio/terra/cli/command/shared/options/ReferencedCloningInstructionsForUpdate.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+   *
+   * <p>This class is meant to be used as a @CommandLine.Mixin.
+   */
+
+  // Can't have a default. This is used for update. Need `null` in case user does not want to
+  // update cloning instructions.
+  @CommandLine.Option(
+      names = "--new-cloning",
+      description =
+          "Instructions for handling when cloning the workspace: COPY_NOTHING, COPY_DEFINITION, COPY_RESOURCE.")
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/command/shared/options/ReferencedCloningInstructionsForUpdate.java:18]: This should be `COPY_NOTHING, COPY_REFERENCE`.

File: src/main/java/bio/terra/cli/serialization/userfacing/input/UpdateReferencedBqTableParams.java [hunk 2]
```diff
@@ -55,6 +59,7 @@ public static class Builder {
     private @Nullable String datasetId;
     private @Nullable String projectId;
     private @Nullable String tableId;
+    private CloningInstructionsEnum cloningInstructions;
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/serialization/userfacing/input/UpdateReferencedBqTableParams.java:62]: mark this `@Nullable` for documentation.

File: src/main/java/bio/terra/cli/serialization/userfacing/input/UpdateReferencedGcsObjectParams.java [hunk 3]
```diff
@@ -24,17 +25,21 @@ public class UpdateReferencedGcsObjectParams {
...
     this.objectName = builder.objectName;
+    this.cloningInstructions = builder.cloningInstructions;
   }
 
   @JsonPOJOBuilder(buildMethodName = "build", withPrefix = "")
   public static class Builder {
     private UpdateResourceParams resourceFields;
     private @Nullable String bucketName;
     private @Nullable String objectName;
+    private CloningInstructionsEnum cloningInstructions;
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/serialization/userfacing/input/UpdateReferencedGcsObjectParams.java:42]: `@Nullable`

File: src/main/java/bio/terra/cli/serialization/userfacing/input/UpdateReferencedGitRepoParams.java [hunk 4]
```diff
@@ -10,16 +11,19 @@
 public class UpdateReferencedGitRepoParams {
   public final UpdateResourceParams resourceFields;
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/serialization/userfacing/input/UpdateReferencedGitRepoParams.java:12]: I think all three of these should be `@Nullable`.

File: src/test/java/unit/BqDatasetReferenced.java [hunk 5]
```diff
@@ -309,60 +309,64 @@ void updateIndividualProperties() throws IOException {
...
             "bq-dataset",
             "--name=" + newName,
-            "--description=" + newDescription);
-    assertEquals(newName, updateDataset.name);
-    assertEquals(newDescription, updateDataset.description);
+            "--description=" + newDescription,
+            "--new-cloning=" + CloningInstructionsEnum.DEFINITION);
+    assertEquals(newName, updatedDataset.name);
+    assertEquals(newDescription, updatedDataset.description);
+    assertEquals(newDescription, updatedDataset.description);
```
Comment (jaycarlton) [inline src/test/java/unit/BqDatasetReferenced.java:344]: Looks like you have this line twice. Also, you're not asserting anything about the cloning instructions. I believe I fixed the bug that required doing a describe before checking it.

File: src/test/java/unit/BqTableReferenced.java [hunk 6]
```diff
@@ -364,66 +364,69 @@ void updateIndividualProperties() throws IOException {
...
         TestCommand.runAndParseCommandExpectSuccess(
             UFBqTable.class,
             "resource",
             "update",
             "bq-table",
             "--name=" + newName,
-            "--description=" + newDescription);
-    assertEquals(newName, updateDataTable.name);
-    assertEquals(newDescription, updateDataTable.description);
+            "--description=" + newDescription,
```
Comment (jaycarlton) [inline src/test/java/unit/BqTableReferenced.java:395]: This probably should be `--new-description`, but for another PR.

File: src/test/java/unit/GcsObjectReferenced.java [hunk 7]
```diff
@@ -441,53 +441,56 @@ void updateIndividualProperties() throws IOException {
...
+    assertEquals(newDescription, updatedBucketObject.description);
 
     // `terra resources describe --name=$newName`
-    describeBucketObject =
+    describedBucketObject =
         TestCommand.runAndParseCommandExpectSuccess(
             UFGcsObject.class, "resource", "describe", "--name=" + newName);
-    assertEquals(newDescription, describeBucketObject.description);
+    assertEquals(newDescription, describedBucketObject.description);
+    assertEquals(CloningInstructionsEnum.REFERENCE, describedBucketObject.cloningInstructions);
```
Comment (jaycarlton) [inline src/test/java/unit/GcsObjectReferenced.java:482]: I thought the new cloning option was `DEFINITION` not `REFERENCE`.

File: src/test/java/unit/GitRepoReferenced.java [hunk 8]
```diff
@@ -201,52 +201,55 @@ void updateIndividualProperties() throws IOException {
...
+    assertEquals(newDescription, updatedGitRepo.description);
 
     // `terra resources describe --name=$newName`
-    describeGitRepo =
+    describedGitRepo =
         TestCommand.runAndParseCommandExpectSuccess(
             UFGitRepo.class, "resource", "describe", "--name=" + newName);
-    assertEquals(newDescription, describeGitRepo.description);
+    assertEquals(newDescription, describedGitRepo.description);
+    assertEquals(CloningInstructionsEnum.REFERENCE, describedGitRepo.cloningInstructions);
```
Comment (jaycarlton) [inline src/test/java/unit/GitRepoReferenced.java:242]: here again I thought the cloning should be `DEFINITION` based on the update command.

**Review by jaycarlton — CHANGES_REQUESTED** [review]

File: src/main/java/bio/terra/cli/command/resource/update/BqTable.java [hunk 9]
```diff
@@ -19,6 +20,7 @@
     description = "Update a BigQuery data table.",
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/resource/update/BqTable.java:20]: nit put on line 27

File: src/test/java/unit/BqDatasetReferenced.java [hunk 10]
```diff
@@ -309,60 +309,64 @@ void updateIndividualProperties() throws IOException {
...
 
     // `terra resources describe --name=$newName`
-    UFBqDataset describeDataset =
+    UFBqDataset describedDataset =
         TestCommand.runAndParseCommandExpectSuccess(
             UFBqDataset.class, "resource", "describe", "--name=" + newName);
-    assertEquals(description, describeDataset.description);
+    assertEquals(description, describedDataset.description);
 
     // update just the description
```
Comment (melissachang) [inline src/test/java/unit/BqDatasetReferenced.java:329]: // update description and cloning instructions

Please do this for all test files:

- Comment is accurate
- terra command in comment is accurate
- Assert cloning instruction was updated

**Review by melissachang — COMMENTED** [review]
Please do some manual testing and paste the output here. This is a good thing to do in general. It tells the reviewers what you changed.

Please include missing args case: `terra resource update git-repo`

File: src/main/java/bio/terra/cli/command/shared/options/ControlledCloningInstructionsForUpdate.java [hunk 11]
```diff
@@ -14,6 +14,8 @@ public class ControlledCloningInstructionsForUpdate {
   // update cloning instructions.
   @CommandLine.Option(
       names = "--new-cloning",
+      // Only COPY_NOTHING, COPY_DEFINITION and COPY_REFERENCE are allowed for referenced resources
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/command/shared/options/ControlledCloningInstructionsForUpdate.java:17]: It's only COPY_NOTHING and COPY_REFERENCE.

File: src/main/java/bio/terra/cli/command/resource/update/BqTable.java (hunk 9 above)
Comment (melissachang) [reply src/main/java/bio/terra/cli/command/resource/update/BqTable.java:20]: Sorry, I mean just move line 23 to line 27

File: src/main/java/bio/terra/cli/command/shared/options/CloningInstructionsForUpdate.java [hunk 12]
```diff
@@ -3,7 +3,7 @@
 import bio.terra.workspace.model.CloningInstructionsEnum;
 import picocli.CommandLine;
 
-public class ControlledCloningInstructionsForUpdate {
+public class CloningInstructionsForUpdate {
   /**
    * Use this --new-cloning flag when updating controlled resources.
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/shared/options/CloningInstructionsForUpdate.java:8]: Change `controlled resources` to `resources`.

File: src/main/java/bio/terra/cli/command/shared/options/CloningInstructionsForUpdate.java [hunk 13]
```diff
@@ -14,8 +14,12 @@ public class ControlledCloningInstructionsForUpdate {
   // update cloning instructions.
   @CommandLine.Option(
       names = "--new-cloning",
+      // Only COPY_NOTHING，COPY_DEFINITION and COPY_RESOURCE are allowed for updating resources
+      // Only COPY_NOTHING and COPY_REFERENCE are allowed for referenced resources
+      // two options or commands in the same class file with same name could not build
```
[the remark this answers is not in the document]
Comment (melissachang) [reply src/main/java/bio/terra/cli/command/shared/options/CloningInstructionsForUpdate.java:19]: Makes sense. Replace 17-22 with:

```
      // For what is allowed, see:
      // https://github.com/DataBiosphere/terra-workspace-manager/blob/main/service/src/main/java/bio/terra/workspace/service/resource/ResourceValidationUtils.java#L448
      description =
          "Instructions for handling when cloning the workspace: COPY_NOTHING, COPY_DEFINITION, or COPY_RESOURCE for controlled resources; COPY_NOTHING or COPY_REFERENCE for referenced resources")
```

File: src/test/java/unit/BqDatasetReferenced.java [hunk 14]
```diff
@@ -309,60 +309,63 @@ void updateIndividualProperties() throws IOException {
...
         TestCommand.runAndParseCommandExpectSuccess(
             UFBqDataset.class, "resource", "describe", "--name=" + newName);
-    assertEquals(description, describeDataset.description);
+    assertEquals(description, describedDataset.description);
 
-    // update just the description
-    // `terra resources update bq-dataset --name=$newName --description=$newDescription`
+    // update description and cloning instructions
+    // `terra resources update bq-dataset --name=$newName --description=$newDescription
+    // --new-cloning=COPY_NOTHING`
```
Comment (melissachang) [inline src/test/java/unit/BqDatasetReferenced.java:331]: Here you have COPY_NOTHING, but on 341 you have NOTHING. For all tests, please make sure your comments are in sync

File: src/test/java/unit/BqDatasetReferenced.java [hunk 15]
```diff
@@ -309,60 +309,63 @@ void updateIndividualProperties() throws IOException {
...
+    assertEquals(newName, updatedDataset.name);
+    assertEquals(description, updatedDataset.description);
 
     // `terra resources describe --name=$newName`
-    UFBqDataset describeDataset =
+    UFBqDataset describedDataset =
         TestCommand.runAndParseCommandExpectSuccess(
             UFBqDataset.class, "resource", "describe", "--name=" + newName);
-    assertEquals(description, describeDataset.description);
+    assertEquals(description, describedDataset.description);
```
Comment (melissachang) [inline src/test/java/unit/BqDatasetReferenced.java:327]: For description:

line 327: Assert old description
line 333: Update new description
line 349: Assert new description

You only have:

line 334: Update cloning instructinos
line 350: Assert new cloning instructions

You're missing:

line 328: assert old cloning instructions

For all we know, update could be broken and cloning instructions could have been set to NOTHING originally. line 328 makes us certain that updating works.

Please do this for all tests.

**Review by melissachang — APPROVED** [review]

File: src/test/java/unit/GitRepoReferenced.java [hunk 16]
```diff
@@ -201,52 +201,56 @@ void updateIndividualProperties() throws IOException {
...
-    assertEquals(newName, updateGitRepo.name);
-    assertEquals(newDescription, updateGitRepo.description);
+            "--description=" + newDescription,
+            "--new-cloning=" + CloningInstructionsEnum.NOTHING);
+    assertEquals(newName, updatedGitRepo.name);
+    assertEquals(newDescription, updatedGitRepo.description);
 
     // `terra resources describe --name=$newName`
-    describeGitRepo =
+    describedGitRepo =
```
Comment (jaycarlton) [inline src/test/java/unit/GitRepoReferenced.java:239]: Same goes here: we shouldn't  need to describe to get the latest values.

File: src/main/java/bio/terra/cli/command/shared/options/CloningInstructionsForUpdate.java [hunk 17]
```diff
@@ -14,8 +14,10 @@ public class ControlledCloningInstructionsForUpdate {
   // update cloning instructions.
   @CommandLine.Option(
       names = "--new-cloning",
+      // For what is allowed, see:
+      // https://github.com/DataBiosphere/terra-workspace-manager/blob/main/service/src/main/java/bio/terra/workspace/service/resource/ResourceValidationUtils.java#L448
       description =
-          "Instructions for handling when cloning the workspace: COPY_NOTHING, COPY_DEFINITION, COPY_RESOURCE.")
+          "Instructions for handling when cloning the workspace: COPY_NOTHING, COPY_DEFINITION, or COPY_RESOURCE for controlled resources; COPY_NOTHING or COPY_REFERENCE for referenced resources")
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/command/shared/options/CloningInstructionsForUpdate.java:20]: Put a period at the end of the description.

**Review by jaycarlton — APPROVED** [review]

**Pull request merged**